### PR TITLE
added `--exclude-tags` option with tags filtering

### PR DESCRIPTION
**Pull request by hitenkoku**
## What Changed

- Added `--exclude-tags` option to `csv-timeline` and `json-timeline` commands to exclude scanning for rules with the specified `tags` field.

I would appreciate it if you could review when you have time.

### CODE REVIEW

**Comment by hitenkoku** [comment]
- display check

```
> ./1118.exe help csv-timeline
Hayabusa v2.7.0 - Dev Build
Yamato Security (https://github.com/Yamato-Security/hayabusa - @SecurityYamato)
Usage:
  hayabusa.exe csv-timeline <INPUT> [OPTIONS]

...
Filtering:
  -E, --EID-filter                Scan only common EIDs for faster speed (./rules/config/target_event_IDs.txt)
  -D, --enable-deprecated-rules   Enable rules with status of deprecated
  -n, --enable-noisy-rules        Enable rules set to noisy (./rules/config/noisy_rules.txt)
  -u, --enable-unsupported-rules  Enable rules with status of unsupported
  -e, --exact-level <LEVEL>       Scan for only specific levels (informational, low, medium, high, critical)
      --exclude-status <STATUS>   Ignore rules according to status (ex: experimental) (ex: stable,test)
      --exclude-tags <TAGS>       Exclude load rules with specific tags (ex: sysmon)
  -m, --min-level <LEVEL>         Minimum level for rules (default: informational)
      --tags <TAGS>               Only load rules with specific tags (ex: attack.execution,attack.discovery)
      --timeline-end <DATE>       End time of the event logs to load (ex: "2022-02-22 23:59:59 +09:00")
      --timeline-start <DATE>     Start time of the event logs to load (ex: "2020-02-22 00:00:00 +09:00")
...
```

- conflict check

```
> ./1118.exe csv-timeline -d ..\hayabusa-sample-evtx\ --tags attack.persistence --exclude-tags hoge
error: the argument '--tags <TAGS>' cannot be used with '--exclude-tags <TAGS>'
Usage: 1118.exe csv-timeline --directory <DIR> --tags <TAGS>

```

- result check

  - this PR (without --exclude-tags)

```
> ./1118.exe csv-timeline -d ..\hayabusa-sample-evtx\ -o 1118-no-extags.csv -C -q --debug
...
Total event log files: 584
Total file size: 137.1 MB

Loading detections rules. Please wait.

Excluded rules: 31
Noisy rules: 12 (Disabled)

Deprecated rules: 165 (4.50%) (Disabled)
Experimental rules: 1973 (53.77%)
Stable rules: 229 (6.24%)
Test rules: 1467 (39.98%)
Unsupported rules: 45 (1.23%) (Disabled)

Hayabusa rules: 159
Sigma rules: 3510
Total enabled detection rules: 3669

Output profile: standard
...
Events with hits / Total events: 19,642 / 47,472 (Data reduction: 27,830 events (58.62%))

Total | Unique detections: 32,228 | 561
Total | Unique critical detections: 56 (0.17%) | 20 (3.57%)
Total | Unique high detections: 6,043 (18.75%) | 245 (43.67%)
Total | Unique medium detections: 1,894 (5.88%) | 184 (32.80%)
Total | Unique low detections: 6,002 (18.62%) | 56 (9.98%)
Total | Unique informational detections: 18,233 (56.58%) | 56 (9.98%)
...
Saved file: 1118-no-extags.csv (32.0 MB)

Elapsed time: 00:00:08.015
Rule Parse Processing Time: 00:00:01.490
Analysis Processing Time: 00:00:06.076
Output Processing Time: 00:00:00.447

Memory usage stats:
heap stats:     peak       total       freed     current        unit       count
  reserved:     2.0 GiB     2.0 GiB     0           2.0 GiB

 committed:   818.1 MiB     1.3 GiB   894.4 MiB   498.8 MiB

     reset:     0
    purged:   425.7 MiB
   touched:    64.2 KiB     1.1 MiB     3.5 GiB    -3.5 GiB
      ok
  segments:    16          19           9          10
      not all freed!
-abandoned:     0           0           0           0
      ok
   -cached:     0           0           0           0
      ok
     pages:     0           0          33.2 Ki    -33.2 Ki
      ok
-abandoned:     0           0           0           0
      ok
 -extended:     0
 -noretire:     0
     mmaps:     0
   commits:     7.2 Ki
    resets:     0
    purges:   300
   threads:    32          32           0          32
      not all freed!
  searches:     0.0 avg
numa nodes:     1
   elapsed:     8.034 s
   process: user: 27.421 s, system: 0.734 s, faults: 265073, rss: 689.1 MiB, commit: 797.3 MiB
```

  - this PR (with --exclude-tags)

```
> ./1118.exe csv-timeline -d ..\hayabusa-sample-evtx\ -o 1118-extags.csv -C -q --debug --exclude-tags sysmon
...
Total event log files: 584
Total file size: 137.1 MB

Loading detections rules. Please wait.

Excluded rules: 82
Noisy rules: 12 (Disabled)

Deprecated rules: 165 (4.56%) (Disabled)
Experimental rules: 1972 (54.51%)
Stable rules: 186 (5.14%)
Test rules: 1460 (40.35%)
Unsupported rules: 45 (1.24%) (Disabled)

Hayabusa rules: 108
Sigma rules: 3510
Total enabled detection rules: 3618

Output profile: standard

....
Results Summary:
Events with hits / Total events: 17,897 / 47,472 (Data reduction: 29,575 events (62.30%))

Total | Unique detections: 29,579 | 527
Total | Unique critical detections: 56 (0.19%) | 20 (3.80%)
Total | Unique high detections: 5,966 (20.17%) | 240 (45.54%)
Total | Unique medium detections: 1,566 (5.29%) | 174 (33.02%)
Total | Unique low detections: 5,724 (19.35%) | 52 (9.87%)
Total | Unique informational detections: 16,267 (55.00%) | 41 (7.78%)

...

Saved file: 1118-extags.csv (29.5 MB)

Elapsed time: 00:00:07.900
Rule Parse Processing Time: 00:00:01.519
Analysis Processing Time: 00:00:05.975
Output Processing Time: 00:00:00.405

Memory usage stats:
heap stats:     peak       total       freed     current        unit       count
  reserved:     2.0 GiB     2.0 GiB     0           2.0 GiB

 committed:   799.2 MiB     1.4 GiB     1.0 GiB   423.6 MiB

     reset:     0
    purged:   505.1 MiB
   touched:    64.2 KiB     1.3 MiB     3.5 GiB    -3.5 GiB
      ok
  segments:    16          21          13           8
      not all freed!
-abandoned:     0           0           0           0
      ok
   -cached:     0           0           0           0
      ok
     pages:     0           0          32.7 Ki    -32.7 Ki
      ok
-abandoned:     0           0           0           0
      ok
 -extended:     0
 -noretire:     0
     mmaps:     0
   commits:     7.7 Ki
    resets:     0
    purges:   311
   threads:    32          32           0          32
      not all freed!
  searches:     0.0 avg
numa nodes:     1
   elapsed:     7.904 s
   process: user: 26.640 s, system: 0.906 s, faults: 266256, rss: 660.7 MiB, commit: 778.3 MiB
```

**Comment by YamatoSecurity** [comment]
@hitenkoku I would like to change the `--tags` naming to `--include-tags` to keep the naming convention uniform. 
Could we update it in this PR or should we create another PR for renaming it?

**Review by YamatoSecurity — APPROVED** [review]
LGTM!

**Comment by hitenkoku** [comment]
Thank you for confirming.
We will make the correction in this Pull Request. Please wait a moment.

> @hitenkoku I would like to change the `--tags` naming to `--include-tags` to keep the naming convention uniform. Could we update it in this PR or should we create another PR for renaming it?

**Comment by hitenkoku** [comment]
```
  ./1118.exe help csv-timeline                                                                                     ─╯
Hayabusa v2.7.0 - Dev Build
Yamato Security (https://github.com/Yamato-Security/hayabusa - @SecurityYamato)

Usage:
  hayabusa.exe csv-timeline <INPUT> [OPTIONS]

Input:
  -d, --directory <DIR>  Directory of multiple .evtx files
  -f, --file <FILE>      File path to one .evtx file
  -l, --live_analysis    Analyze the local C:\Windows\System32\winevt\Logs folder
  -J, --JSON-input       Scan JSON formatted logs instead of .evtx (.json or .jsonl)

Output:
  -G, --GeoIP <MAXMIND-DB-DIR>  Add GeoIP (ASN, city, country) info to IP addresses
  -H, --HTML-report <FILE>      Save Results Summary details to an HTML report (ex: results.html)
  -M, --multiline               Output event field information in multiple rows
  -o, --output <FILE>           Save the timeline in CSV format (ex: results.csv)
  -p, --profile <PROFILE>       Specify output profile
  -R, --remove-duplicate-data   Duplicate field data will be replaced with "DUP"

Display Settings:
      --no-color            Disable color output
  -N, --no-summary          Do not display Results Summary (slightly faster speed)
  -q, --quiet               Quiet mode: do not display the launch banner
  -v, --verbose             Output verbose information
  -T, --visualize-timeline  Output event frequency timeline (terminal needs to support unicode)

Filtering:
  -E, --EID-filter                Scan only common EIDs for faster speed (./rules/config/target_event_IDs.txt)
  -D, --enable-deprecated-rules   Enable rules with status of deprecated
  -n, --enable-noisy-rules        Enable rules set to noisy (./rules/config/noisy_rules.txt)
  -u, --enable-unsupported-rules  Enable rules with status of unsupported
  -e, --exact-level <LEVEL>       Scan for only specific levels (informational, low, medium, high, critical)
      --exclude-status <STATUS>   Ignore rules according to status (ex: experimental) (ex: stable,test)
      --exclude-tags <TAGS>       Exclude load rules with specific tags (ex: sysmon)
      --include_tags <TAGS>       Only load rules with specific tags (ex: attack.execution,attack.discovery)
  -m, --min-level <LEVEL>         Minimum level for rules (default: informational)
      --timeline-end <DATE>       End time of the event logs to load (ex: "2022-02-22 23:59:59 +09:00")
      --timeline-start <DATE>     Start time of the event logs to load (ex: "2020-02-22 00:00:00 +09:00")

General Options:
  -C, --clobber                          Overwrite results files
  -Q, --quiet-errors                     Quiet errors mode: do not save error logs
  -r, --rules <DIR/FILE>                 Specify a custom rule directory or file (default: ./rules)
  -c, --rules-config <DIR>               Specify custom rule config directory (default: ./rules/config)
      --target-file-ext <EVTX_FILE_EXT>  Specify additional file extensions (ex: evtx_data) (ex: evtx1,evtx2)
  -t, --threads <NUMBER>                 Number of threads (default: optimal number for performance)

Time Format:
      --European-time     Output timestamp in European time format (ex: 22-02-2022 22:00:00.123 +02:00)
      --ISO-8601          Output timestamp in ISO-8601 format (ex: 2022-02-22T10:10:10.1234567Z) (Always UTC)
      --RFC-2822          Output timestamp in RFC 2822 format (ex: Fri, 22 Feb 2022 22:00:00 -0600)
      --RFC-3339          Output timestamp in RFC 3339 format (ex: 2022-02-22 22:00:00.123456-06:00)
      --US-military-time  Output timestamp in US military time format (ex: 02-22-2022 22:00:00.123 -06:00)
      --US-time           Output timestamp in US time format (ex: 02-22-2022 10:00:00.123 PM -06:00)
  -U, --UTC               Output time in UTC format (default: local time)
```

File: src/detections/configs.rs [hunk 1]
```diff
@@ -1018,8 +1018,8 @@ pub struct OutputOption {
     pub exclude_status: Option<Vec<String>>,
 
     /// Only load rules with specific tags (ex: attack.execution,attack.discovery)
-    #[arg(help_heading = Some("Filtering"), long = "tags", value_name = "TAGS", use_value_delimiter = true, value_delimiter = ',', display_order = 460)]
-    pub tags: Option<Vec<String>>,
+    #[arg(help_heading = Some("Filtering"), long = "include_tags", value_name = "TAGS", conflicts_with = "exclude_tags", use_value_delimiter = true, value_delimiter = ',', display_order = 351)]
```
Comment (fukusuket) [inline src/detections/configs.rs:1021]:
```suggestion
    #[arg(help_heading = Some("Filtering"), long = "include-tags", value_name = "TAGS", conflicts_with = "exclude-tags", use_value_delimiter = true, value_delimiter = ',', display_order = 351)]
```

**Review by fukusuket — CHANGES_REQUESTED** [review]
Would a `-`(hyphen) be correct instead of an `underscore`?
(Sorry if I misunderstood!)

File: src/detections/configs.rs (hunk 1 above)
Comment (fukusuket) [reply src/detections/configs.rs:1021]: @hitenkoku 
Sorry for my misunderstanding!🙇 Where is the `_`(underscore) below defined?
```
      --exclude-tags <TAGS>       Exclude load rules with specific tags (ex: sysmon)
      --include_tags <TAGS>       Only load rules with specific tags (ex: attack.execution,attack.discovery)
```

Comment (hitenkoku) [reply src/detections/configs.rs:1021]: I tried to modify it as you suggested and got it to work, but I made it set both items specified in conflict_with, but I confirmed that the analysis starts as it is without any control.

image

Comment (fukusuket) [inline src/detections/configs.rs:1021]:
```suggestion
    #[arg(help_heading = Some("Filtering"), long = "include-tags", value_name = "TAGS", conflicts_with = "exclude_tags", use_value_delimiter = true, value_delimiter = ',', display_order = 351)]
```

Comment (fukusuket) [reply src/detections/configs.rs:1021]: @hitenkoku 
Thank you for checking!
How about changing only this? (does not change `conflicts_with`)

Comment (hitenkoku) [reply src/detections/configs.rs]: @fukusuket Thank you for your suggest.
I apologize for the incorrect description on my part.
The content is indeed correct. I have adopted your suggestion.

**Review by fukusuket — APPROVED** [review]
@hitenkoku 
Thank you for quick fix :) (and I'm sorry for my wrong suggestion🙇) I confirmed as follows,  LGTM!!🚀
- No unnecessary diffs compared to v2.6.0
- No unnecessary diffs compared to After merging the results of `--include-tags` and `--exclude-tags`

**Comment by hitenkoku** [comment]
@YamatoSecurity @fukusuket  Thanks for your quick review. I will merge it,

**Pull request merged**